### PR TITLE
Updating Cards to have checks for frozen heap

### LINKED ISSUE #5275: Blacklist chat log is not correct with Boomerang
wlip started their turn 8 with 11  and 5 cards in their Grip.
wlip uses Beth Kilrain-Chang to draw 1 card.
wlip uses Security Testing to target Archives.
wlip spends  and pays 2  to install Boomerang.
wlip uses Boomerang to target ICE protecting HQ at position 0.
wlip spends  and pays 1  to play Diversion of Funds.
wlip approaches ICE protecting HQ at position 0.
The_Greasy_Thumb pays 2  to rez IP Block.
wlip encounters IP Block protecting HQ at position 0.
wlip to use Boomerang to break 2 subroutines on IP Block (" Trace 3 - Give the Runner 1 tag" and " End the run if the Runner is tagged").
wlip has no further action.
wlip passes IP Block protecting HQ at position 0.
wlip approaches HQ.
The_Greasy_Thumb pays 0  to rez Blacklist.
wlip uses Paragon to gain 1  and look at the top card of Stack.
wlip uses Paragon to add the top card of Stack to the bottom.
wlip chooses to use the replacement effect from Diversion of Funds.
wlip uses Diversion of Funds to force the Corp to lose 5, and gain 5.
wlip places a power counter on The Turning Wheel.
wlip uses Boomerang to shuffle a copy of Boomerang back into the Stack.
wlip spends  to make a run on Server 5.
wlip approaches ICE protecting Server 5 at position 1.

Boomerang correctly does not leave the heap, but I think we need a message in the chat log to say it was prevented.

### PR DESCRIPTION
**Pull request by tblackwe**
Closes #5275

### CODE REVIEW

**Comment by tblackwe** [comment]
needs work, dont merge

**Comment by tblackwe** [comment]
actually, code was fine. my knowledge of the rules of the game was broken.

File: test/clj/game/cards/hardware_test.clj [hunk 1]
```diff
@@ -860,7 +875,23 @@
...
+      (take-credits state :corp)
+      (trash-from-hand state :runner "Datasucker")
+      (rez state :corp (refresh (get-content state :remote1 0)))
+      (is (= 1 (count (:discard (get-runner)))) "Datasucker in heap")
+      (play-from-hand state :runner "Clone Chip")
+      (let [chip (get-hardware state 0)]
+        (card-ability state :runner chip 0)
+        (is (empty? (:prompt (get-runner))) "Clone Chip Prompt did not came up")
+        (is (= 1 (count (:discard (get-runner)))) "Datasucker in still in heap"))))
+  )
```
Comment (NoahTheDuke) [inline test/clj/game/cards/hardware_test.clj:893]: :eyes:

File: test/clj/game/cards/hardware_test.clj [hunk 2]
```diff
@@ -3359,7 +3390,31 @@
...
+        (click-prompt state :runner "Pay to steal")
+        (click-card state :runner "Corroder")
+        (card-ability state :runner (get-hardware state 0) 0)
+        (is (not (empty? (:prompt (get-runner)))) "Simulchip prompt came up")
+        (click-prompt state :runner "Done")
+        (rez state :corp (refresh (get-content state :remote2 0)))
+        (play-from-hand state :runner "Simulchip")
+        (card-ability state :runner (get-hardware state 0) 0)
+        (is (empty? (:prompt (get-runner))) "Simulchip prompt did not come up")
+        )))
```
Comment (NoahTheDuke) [inline test/clj/game/cards/hardware_test.clj:3417]: 👀

File: test/clj/game/cards/hardware_test.clj [hunk 3]
```diff
@@ -4017,3 +4072,49 @@
...
+        (rez state :corp (refresh wrap))
+        (run-continue state)
+        (card-ability state :runner brang 0)
+        (click-prompt state :runner "End the run")
+        (run-continue state :approach-server)
+        (run-continue state :access-server)
+        (click-prompt state :runner "No action")
+        (click-prompt state :runner "Yes")
+        (is (= 0 (count (:discard (get-runner)))) "Boomerang in stack")
+        (last-log-contains? state "Runner uses Boomerang to shuffle a copy of Boomerang back into the Stack."))))
```
Comment (NoahTheDuke) [inline test/clj/game/cards/hardware_test.clj:4097]: Needs to be wrapped in an `is` for this to act as a test.

File: test/clj/game/cards/hardware_test.clj [hunk 4]
```diff
@@ -630,6 +630,20 @@
...
+                          :deck ["Stimhack" "Corroder" "Yog.0" "Mimic"]
+                          :discard ["Stimhack"]}})
+      (play-from-hand state :corp "Blacklist" "New remote")
+      (take-credits state :corp)
+      (play-from-hand state :runner "Buffer Drive")
+      (rez state :corp (refresh (get-content state :remote1 0)))
+      (let [bufferdrive (get-hardware state 0)]
+        (card-ability state :runner bufferdrive 0)
+        (is (empty? (:prompt (get-runner))) "Buffer Drive Prompt did not came up"))
+      ))
```
Comment (NoahTheDuke) [inline test/clj/game/cards/hardware_test.clj:646]: 👀

File: src/clj/game/cards/hardware.clj [hunk 5]
```diff
@@ -189,7 +189,7 @@
                                       :once :per-run
                                       :duration :end-of-run
                                       :optional
-                                      {:req (req (and (:successful target)))
+                                      {:req (req (and (:successful target) (not (zone-locked? state :runner :discard))))
```
Comment (NoahTheDuke) [inline src/clj/game/cards/hardware.clj:192]: Let's split this onto a new line.

**Comment by NoahTheDuke** [comment]
Looks great! Keep up the good work.

File: test/clj/game/cards/hardware_test.clj (hunk 1 above)
Comment (tblackwe) [reply test/clj/game/cards/hardware_test.clj:893]: Are the eyes a request to fix parens new lines? or something else?

Comment (NoahTheDuke) [reply test/clj/game/cards/hardware_test.clj]: Yep! just a silly way of noting that they're there.

File: src/clj/game/cards/events.clj [hunk 6]
```diff
@@ -609,7 +609,8 @@
              :effect (effect (access-bonus :rd (max 0 (min 4 (available-mu state)))))}]})
 
 (defcard "Déjà Vu"
-  {:prompt "Choose a card to add to Grip" :choices (req (cancellable (:discard runner) :sorted))
+  {:req (req (not (zone-locked? state :runner :discard)))
+   :prompt "Choose a card to add to Grip" :choices (req (cancellable (:discard runner) :sorted))
```
Comment (NoahTheDuke) [inline src/clj/game/cards/events.clj:613]: mind splitting this onto two lines for me? it's a hold over from very old code, and i'm not a fan.

File: src/clj/game/cards/events.clj [hunk 7]
```diff
@@ -1733,11 +1734,13 @@
                                :effect (effect (unregister-floating-events :until-runner-turn-begins))}]))}}})
 
 (defcard "Levy AR Lab Access"
-  {:msg "shuffle their Grip and Heap into their Stack and draw 5 cards"
-   :rfg-instead-of-trashing true
-   :async true
-   :effect (effect (shuffle-into-deck :hand :discard)
-                   (draw eid 5 nil))})
+         {:msg                     (msg (if (not (zone-locked? state :runner :discard))
```
Comment (NoahTheDuke) [inline src/clj/game/cards/events.clj:1737]: let's keep the spacing between keys and values to 1 space

File: test/clj/game/cards/events_test.clj [hunk 8]
```diff
@@ -4484,10 +4603,10 @@
       (is (nil? (get-run)) "Run is ended")))
   (testing "with Gauntlet #2942"
     (do-game
-      (new-game {:corp {:deck [(qty "Hedge Fund" 5)]
-                        :hand [(qty "Vanilla" 5)]
-                        :credits 10}
-                 :runner {:hand ["The Gauntlet" "Corroder" "Rip Deal"]
+      (new-game {:corp   {:deck    [(qty "Hedge Fund" 5)]
```
Comment (NoahTheDuke) [inline test/clj/game/cards/events_test.clj:4606]: same with the spacing here

Comment (tblackwe) [reply test/clj/game/cards/events_test.clj:4606]: appears I need to figure out how to edit the cursive ide autoformatter.

File: test/clj/game/cards/events_test.clj [hunk 9]
```diff
@@ -2460,12 +2461,12 @@
...
-      (new-game {:runner {:hand ["Harmony AR Therapy"]}})
-      (take-credits state :corp)
-      (play-from-hand state :runner "Harmony AR Therapy")
-      (is (empty? (:prompt (get-runner))) "HART was not played")))
+  ;(testing "Can play with empty heap"
+  ;  (do-game
+  ;    (new-game {:runner {:hand ["Harmony AR Therapy"]}})
+  ;    (take-credits state :corp)
+  ;    (play-from-hand state :runner "Harmony AR Therapy")
+  ;    (is (= 2 (count (prompt-buttons :runner))) "No Cards to Shuffle")))
```
Comment (NoahTheDuke) [inline test/clj/game/cards/events_test.clj:2469]: Did you mean to remove this test?

File: test/clj/game/cards/events_test.clj [hunk 10]
```diff
@@ -5211,7 +5231,21 @@
...
+                          :deck ["Pelangi"]
+                          :discard ["Morning Star"]}})
+      (play-from-hand state :corp "Blacklist" "New remote")
+      (rez state :corp (refresh (get-content state :remote1 0)))
+      (take-credits state :corp)
+      (play-from-hand state :runner "Test Run")
+      (is (= "Install a program from your Stack?" (:msg (prompt-map :runner))) "Stack is only option")
+      (is (= 1 (-> (prompt-map :runner) :choices count)) "Runner has 1 choice")
+      (is (= ["Stack"] (prompt-buttons :runner)) "Runner's only choice is Stack"))))
+
```
Comment (NoahTheDuke) [inline test/clj/game/cards/events_test.clj:5248]: Extra space here.

File: test/clj/game/cards/programs_test.clj [hunk 11]
```diff
@@ -4579,7 +4579,7 @@
         (is (= 2 (get-strength (refresh corr))) "Corroder returned to normal strength")))))
 
 (deftest trope
-  ;; Trope
+  ;; Trope TODO FIX ME
```
Comment (NoahTheDuke) [inline test/clj/game/cards/programs_test.clj:4582]: hahaha

File: src/clj/game/cards/events.clj [hunk 12]
```diff
@@ -1268,10 +1268,14 @@
...
     {:async true
-     :req (req (pos? (count (:discard runner))))
      :rfg-instead-of-trashing true
-     :effect (req (show-wait-prompt state :corp (str "Runner to resolve " (:title card)))
-               (continue-ability state side (choose-next '() nil (sort (distinct (map :title (:discard runner))))) card nil))}))
+     :effect (req (if (and (not (zone-locked? state :runner :discard))
+                           (pos? (count (:discard runner))))
+                    (do (show-wait-prompt state :corp (str "Runner to resolve " (:title card)))
+                        (continue-ability state side (choose-next '() nil (sort (distinct (map :title (:discard runner))))) card nil))
+                    (do (system-msg state :runner (str "used " (:title card) " to shuffle their Stack"))
```
Comment (NoahTheDuke) [inline src/clj/game/cards/events.clj:1276]:
```suggestion
                    (do (system-msg state :runner (str "uses " (:title card) " to shuffle their Stack"))
```

File: src/clj/game/cards/events.clj [hunk 13]
```diff
@@ -1635,30 +1639,36 @@
...
-                                           :card #(and (runner? %)
-                                                       (in-discard? %))}
-                                 :effect (req (doseq [c targets]
-                                                (move state side c :deck))
-                                              (system-msg state :runner (str "shuffles " (string/join ", " (map :title targets))
-                                                                             " from their Heap into their Stack, and draws 1 card"))
-                                              (shuffle! state :runner :deck)
-                                              (draw state :runner eid 1 nil))}
-                                card nil)))))})
+         {:req (req (pos? (+ (count (:deck runner)) (count (:discard runner)))))
```
Comment (NoahTheDuke) [inline src/clj/game/cards/events.clj:1642]: indentation has gotten messed up here

File: test/clj/game/cards/events_test.clj (hunk 9 above)
Comment (tblackwe) [reply test/clj/game/cards/events_test.clj]: older commit, got fixed but the file is currently in a conflicted merge. thanks!

**Comment by tblackwe** [comment]
alright @NoahTheDuke PR is open. feel free to stop being nice.

File: test/clj/game/cards/events_test.clj [hunk 14]
```diff
@@ -4553,8 +4553,9 @@
       (rez state :corp (refresh (get-content state :remote1 0)))
       (take-credits state :corp)
       (trash-from-hand state :runner "Morning Star")
-      (play-from-hand state :runner "Retrieval Run")
-      (is (not (:run @state)) "Retrieval run didn't get played")))
+      (play-run-event state "Retrieval Run" :archives)
+      (click-prompt state :runner "Retrieval Run")
+      (is (not (:run @state)) "Retrieval run didn't attempt to install from heap.")))
```
Comment (NoahTheDuke) [inline test/clj/game/cards/events_test.clj:4558]: Maybe include an empty prompt check here too? Just to be clear.

**Pull request merged**